### ISSUE TITLE
GPU workload too small (even with large data, large model)

### ISSUE THREAD

**Issue by fordicus** [microsoft/LightGBM#2658]
## Environment info

Operating System:
  Windows 10 x64

CPU/GPU model: 
  CPU: AMD Ryzen 5 2400G 3.6GHz
  GPU: GTX 1080 Ti 
  Memory: 11 [GB] on GPU, 64 [GB] on Motherboard

C++/Python/R version:
  Python 3.7.3

LightGBM version or commit hash:
  LightGBM 2.3.2

## Error message

It is not a printed error. But let me explain. There is initial prints using verbosity(1) below. 
As one can see, **input data is not a small one. 
That is 10 [GB} saved in double and there is no sparsity.**

The symptom is that the exploitation of GPU observed on Task Manager (Windows) is too small. There has been a suggestion, probably it can happen if the model under training has a small figure. I tried with **sufficiently large model;
{
num_boost_round = 10000	#1e4
'num_leaves': 		10000,	#1e4
'max_depth': 		-1,
'max_bin':			255
},**
but still, 
**GPU workload is roughly 15%, and 
GPU memory load is only 1.6 [GB] out of total 11 [GB] available.** 

I will append the code snippet, too, so that one can have a feeling more easily on this matter. **I am willing to help improving this scope, so please guide me which I should check first.**

----------------------------------------------------------------------------------------
prints using verbosity(1)
----------------------------------------------------------------------------------------
[LightGBM] [Info] This is the **GPU** trainer!!
[LightGBM] [Info] Total Bins 38250
[LightGBM] [Info] Number of data points in the **train set: 6666443, number of used features: 150**
[LightGBM] [Info] Using requested OpenCL platform 0 device 0
[LightGBM] [Info] Using GPU Device: **GeForce GTX 1080 Ti**, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 12
[LightGBM] [Info] 150 dense feature groups **(966.36 MB)** transferred to GPU in 1.353269 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score -0.000738

## Reproducible examples

lgb_params = {
	'verbosity':		1,
	'num_leaves': 		10000,	#1e4
	'max_depth': 		-1,
	'shrinkage_rate':	0.01,
	'boosting_type': 	'gbdt',
	'objective':		'regression',
	'metric':			'mse',
	'num_threads':		1,		# reduce cpu load
	'device_type':		'gpu',
	'gpu_platform_id':	0,
	'gpu_device_id':	0,
	'gpu_use_dp':		False,
	'max_bin':			255
}

xtrain, xvalid, ytrain, yvalid = train_test_split(df.iloc[:,:-1], df.iloc[:,-1]) #pandas DataFile
lgb_estimator = None

lgb_estimator = lgb.train(
	lgb_params,
	num_boost_round=my_num_trees,
	init_model=lgb_estimator,				# pass partially trained model (later)
	keep_training_booster=True, 				# for incremental learning (later)
	train_set=lgb.Dataset(xtrain, ytrain),
	valid_sets=[lgb.Dataset(xvalid, yvalid)],
	valid_names=["Valid"],
	verbose_eval=100)

## Steps to reproduce

1. I've installed in this way.
>> git clone --recursive https://github.com/Microsoft/LightGBM.git
>> python setup.py install --gpu --boost-include-dir=C:\library\boost_1_72_0

2. The object decreases anyway, and I don't see any warning or an error when I run a small python code to test LightGBM.

3. The model setting **{
num_boost_round = 10000	#1e4
'num_leaves': 		10000,	#1e4
'max_depth': 		-1,
'max_bin':			255
},** is not small at all I think. But as I am a newbie to this literature, please give me a guideline how to locate the troublesome part and how to improve. Thanks.

**GPU workload is roughly 15%, and 
GPU memory load is only 1.6 [GB] out of total 11 [GB] available.**

**Comment by fordicus** [comment]
Full code is here.


from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import lightgbm as lgb
import pandas as pd
import numpy as np
import h5py
import time
import gc


print('')
gc.enable()
gc.collect()

data_pne = "E:\\Data\\0000000000.mat"
print(">> loading", data_pne)
start_time = time.time()
mat = h5py.File(data_pne, 'r')
df = pd.DataFrame(np.vstack((mat['X'], mat['y'])))
df = df.transpose()
mat.close()
gc.collect()
print(">> loading data took: {:.2f} [sec]".format(time.time() - start_time))

my_num_trees = 			10000	#1e4

lgb_params = {
	'verbosity':		1,
	'num_leaves': 		10000,	#1e4
	'max_depth': 		-1,
	'shrinkage_rate':	0.01,
	'boosting_type': 	'gbdt',
	'objective':		'regression',
	'metric':			'mse',
	'num_threads':		1,
	'device_type':		'gpu',
	'gpu_platform_id':	0,
	'gpu_device_id':	0,
	'gpu_use_dp':		False,
	'max_bin':			255
}

lgb_estimator = None

print("\n>> spliting starts")
start_time = time.time()
xtrain, xvalid, ytrain, yvalid = train_test_split(df.iloc[:,:-1], df.iloc[:,-1])
print(">> spliting took: {:.2f} [sec]".format(time.time() - start_time))
print("\n>> training starts")
start_time = time.time()
lgb_estimator = lgb.train(
	lgb_params,
	num_boost_round=my_num_trees,
	init_model=lgb_estimator,					# pass partially trained model
	keep_training_booster=True, 				# for incremental learning
	train_set=lgb.Dataset(xtrain, ytrain),
	valid_sets=[lgb.Dataset(xvalid, yvalid)],
	valid_names=["Valid"],
	verbose_eval=100) 							# print frequency of metric

del xtrain, xvalid, ytrain, yvalid
gc.collect()
print(">> training took: {:.2f} [sec]".format(time.time() - start_time))


""" performance of predictions """
trues = df.iloc[test_head:test_tail,-1]
print(">> start prediction")
start_time = time.time()
preds = lgb_estimator.predict(df.iloc[:,:-1])
print(">> prediction took: {:.2f} [sec]".format(time.time() - start_time))

plt.scatter(preds, trues)
plt.show()


""" single prediction runtime """
a_random_sample_idx = np.random.randint(low=0, high=df.values.shape[0], size=1)
a_random_sample = df.iloc[a_random_sample_idx,:-1]
start_time = time.time()
pred = lgb_estimator.predict(a_random_sample)
print(">> single prediction took: {:.2f} [sec]".format(time.time() - start_time))
true = df.iloc[a_random_sample_idx,-1]
print("(pred, true) =", pred, true,"\n")

**Comment by guolinke** [comment]
as GPU is used to construct feature histograms, which is bounded by memory, not the computation. It is expected to use the small GPU workload.

**Comment by fordicus** [comment]
> as GPU is used to construct feature histograms, which is bounded by memory, not the computation. It is expected to use the small GPU workload.

I was not particularly interested in the internals of an approach at the moment. But it sounds like, "exploitation of GPU computing power cannot much improve from the current implementation of LightGBM due to the nature of the algorithm". Am I correct?

**Comment by guolinke** [comment]
@fordicus I think it mainly bounded by the memory speed. refer to https://arxiv.org/pdf/1706.08359.pdf
